server/utils/pbench_server.py

```diff
@@ -0,0 +1,136 @@
+        filename = request.headers.get("filename")
+        tarball_md5 = f"{filename}.md5.check"
+
+        if not os.path.exists(filename):
+            logger.debug("Filename %s does not exist", filename)
+            abort(400, message='Filename does not appear to be a valid one')
+        if not os.path.exists(tarball_md5):
+            logger.debug("MD5 check sum does not exist for %s", filename)
+            abort(400, message='Missing MD5 check sum')
+
+        try:
+            hash_md5 = hashlib.md5()
+            with open(filename, "rb") as tar:
+                for chunk in iter(lambda: tar.read(4096), b""):
+                    hash_md5.update(chunk)
+
+                with open(tarball_md5, "rb") as check_sum:
+                    hash_check = check_sum.read()
+                    if hash_md5.hexdigest() == hash_check.decode("utf-8"):
+                        result = True
+        except Exception:
+            logger.error("md5sum failed for %s, skipping", filename)
+            if os.path.exists(filename):
+                os.remove(filename)
+            if os.path.exists(tarball_md5):
+                os.remove(tarball_md5)
+            abort(400, message=f"md5sum check failed for {filename}")
+
+        if result:
+            resp = jsonify(dict(result='md5sum check is correct'))
+            resp.status_code = 200
+            return resp
+        else:
+            abort(400, message=f"md5sum check failed for {filename}")
+
+
+class Upload(Resource):
+    def post(self):
+        if not request.headers.get("filename"):
+            logger.debug("Missed filename in header")
+            abort(400, message='Missing filename header in request')
+        filename = secure_filename(request.headers.get("filename"))
+        logger.debug(f"Receiving file: {filename}")
+        if not allowed_file(filename):
+            logger.debug("Bad file extension received")
+            abort(400, message='File extension not supported. Only .xz')
+        upload_directory = config.get("pbench-receive-dir-prefix")
+        if not upload_directory:
+            logger.debug("Missing config variable for pbench-receive-dir-prefix")
+            abort(400, message="Missing config variable for pbench-receive-dir-prefix")
+        if not os.path.exists(upload_directory):
+            os.mkdir(upload_directory)
+        full_path = os.path.join(upload_directory, filename)
+        with open(full_path, "wb") as f:
+            chunk_size = 4096
+            logger.debug("Writing chunks")
+
+            while True:
+                chunk = request.stream.read(chunk_size)
+                if len(chunk) == 0:
+                    break
+
+                f.write(chunk)
+
+        resp = jsonify(dict(message='File successfully uploaded'))
+        resp.status_code = 201
+        return resp
+
+
+api.add_resource(Upload, f"{app.config['REST_URI']}/upload")
+api.add_resource(Check, f"{app.config['REST_URI']}/check")
```

Comment:
Can we merge the `Check` method into the `Upload` method by having the `Upload` method require a header that contains a MD5 value that can be used to verify the uploaded data?

Reply:
I'd prefer to keep this 2 separated, the hash check is a lengthy process as well as the upload, would be more prone to timeouts, plus I prefer the granularity of knowing what part of the upload went wrong.